Cython/Compiler/Code.py

```diff
@@ -2188,44 +2190,68 @@ def as_pyobject(self, cname, type):
         return typecast(py_object_type, type, cname)
 
     def put_gotref(self, cname, type):
+        if type.refcounting_tracked_by_refnanny and self.funcstate:
+            self.funcstate.needs_refnanny = True
```

Comment:
Wow, this code multiplication is awful. Why not at least move this into `PyObjectType`? In the methods there, you could just always say `code.funcstate.needs_refnanny = nanny/True`, without further conditions.

That said, if these refcounting code bits are really the only point where we can take this decision, then the granularity is ok. But the question remains if we need to decide it here, at the very bottom of the code generation. Isn't it enough to check the types of named variables and temp variables after generating the function? Or, if we have local (parallel?) temp scoping for some reason, let the code funcstate remember whether it created any Python object temps along the way, and check for that flag.

You probably put a lot of thoughts and effort into this, so, could you explain why keeping the decision in the function generation code (rather than the low-level refcounting code) can't work?

Reply:
> Wow, this code multiplication is awful. Why not at least move this into `PyObjectType`? In the methods there, you could just always say `code.funcstate.needs_refnanny = nanny/True`, without further conditions.

That would be better - I'd missed that as an option. I'll do that

>  But the question remains if we need to decide it here, at the very bottom of the code generation. Isn't it enough to check the types of named variables and temp variables after generating the function? Or, if we have local (parallel?) temp scoping for some reason, let the code funcstate remember whether it created any Python object temps along the way, and check for that flag.

That's what we did before, and it mostly works. I think we'd mainly need to add a flag that things creating local Python variables can set. Where our current code fails a little is for Python object arguments that aren't reassigned (e.g. `self` quite frequently) which don't need reference counting, but get considered like they do. The nice thing about just tracking all decref/gotref/incref/etc is that it should give the right answer.